Trigger a notification at 8pm if no contribution

KeziahMoselle opened the pull request:
I'd like to be warned when I didn't commited this day so I thought it would be cool to add this feature.
I added 2 dependencies : [cron](https://github.com/kelektiv/node-cron) and [node-notifier](https://github.com/mikaelbr/node-notifier).
`cron` to schedule the notification at 20pm and `node-notifier` to have a cross-platform notification.

So I added a preference to enable or disable it : (Disabled by default in the store)
![preview.png](https://puu.sh/BFwtV/a766bfd07e.png)

When the notification trigger :
![preview.png](https://puu.sh/BFwvZ/e4d65f1a9d.png)

So should I edit the icon black/red to have a white/red one ?
And in the distributed app `node-notifier` can't find the icon path, if you have an idea on how I can fix that ?
![preview.png](https://puu.sh/BFwBh/5821b6c8e2.png)

There is another thing, when the job is triggered I wanted to launch `requestContributionData()` before the check of `if (!store.get('contributedToday'))` because if we don't check just before and someone commit between the last 15min and 20pm it will trigger the notification. I tried to create an async function and await on the function but it doesn't work.

Commit:
Add notifications / tray setting

Commit:
Trigger notification only if contributedToday = false

Commit:
notification = false by default

Commit:
Remove debug lines

Commit:
Merge branch 'master' into feature-notification-when-no-activity

Commit:
Check currentStreak before notify

KeziahMoselle commented:
Now, it will execute a simple check before notify at 20pm

Commit:
Add white icon for win32 platform (In notification)

KeziahMoselle commented:
On Windows the notification icon looks like that now : 
![preview.png](https://puu.sh/BGLPG/efc4ca21e7.png)

Commit:
Remove path require

jamieweavis commented:
Hey @KeziahMoselle, thanks for another pull request!

Electron [supports notifications](https://electronjs.org/docs/api/notification) out of the box - we shouldn't need to use [node-notifier](https://github.com/mikaelbr/node-notifier) for this, I think that library is intended to be used outside of the Electron eco-system.

If we could remove that extra dependency in favour of Electrons own notifications that would be great!

jamieweavis reviewed (changes requested):
Hey @KeziahMoselle, thanks for another pull request!

Electron [supports notifications](https://electronjs.org/docs/api/notification) out of the box - we shouldn't need to use [node-notifier](https://github.com/mikaelbr/node-notifier) for this, I think that library is intended to be used outside of the Electron eco-system.

If we could remove that extra dependency in favour of Electrons own notifications that would be great!

Commit:
Replace node-notifier by Notification Electron

KeziahMoselle commented:
That's true, I replaced it.
And now in the distributed executable the icon is showed correctly.
![preview.png](https://puu.sh/BLo1p/5a20415bb3.png)
(Don't mind the hour, I changed the cronTime to debug)

Commit:
Remove unecessary code

Commit:
Replace ` by '

Commit:
Remove unecessary code

yursky commented:
Hope it's alright to come in and give some input on this PR. But it would be awesome to be able to configure the time at which the notification occurs. 

IMO this would work best through some kind of preferences pane, or a popup dropdown similar to what is currently done for the github username. (not sure if this fits within the scope of #11) 

That being said, I'm not sure what @jamieweavis has in mind for this project, whether they want to keep it as minimal as possible, or if continuing to expand or add features is something they're interested in. Would be cool to hear your vision @jamieweavis 😄

jamieweavis commented:
I was about to suggest the same thing @yursky! 😄

I've thought about adding this feature before and I came to the conclusion that the best solution would be converting the username window to a general preferences window where both options, and future options can go.

What do you think @KeziahMoselle? I know this more work, and as mentioned it might be out of scope of this issue.

Regarding my vision for the project @yursky, I'm happy to continue adding features as long as they add value! Feel free to give input, submit PRs, open issues etc... As contributors both of your input is extremely valuable!

KeziahMoselle commented:
This is a great idea !

I will add the ability to change the hour of the notification (with a popup similar to the username window for now).

I don't know if you want me to add the preferences window too ? I can do it but if you want a specific UI ?

jamieweavis commented:
If you want to have a go at creating the preferences window then go for it! I would suggest maybe using a UI framework now that we require more interface elements, perhaps [bulma](https://bulma.io/)?

The preferences window is probably out of scope of this issue, if you don't want to do it then feel free to do another window similar to the username one, but remember that this may be removed in the future in favour of a preferences window.

KeziahMoselle commented:
I want to try adding it, but I think I will create an another PR to replace the username and hour popups. In this PR I will just add the ability to change the hour of the notification !
I hope it's okay for you !

Commit:
Add notification popup & Set time to the store

Commit:
Set time when updated

KeziahMoselle commented:
So in theory it works, but it doesn't and I don't know why, if anyone has an idea

yursky commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+  function setNotificationTime(event, time) {
+    notificationWindow.close();
+    if (time && time !== store.get('notification.time')) {
+      let hours = time.split(':')[0];
```
These can both be `const`

yursky commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+
+  const job = new CronJob({
+    cronTime: `0 0 20 * * *`,
+    onTick: async function() {
```
We can rewrite this function header as `async onTick() {` since we have babel goodness to match airbnb style guide

KeziahMoselle replied:
I'm sorry I don't follow you here, can you provide me an example ?
I'm suppose to write : 
```js
onTick: async onTick() {}
```
What's the point with Babel and Airbnb style guide ?

yursky replied:
Right, the point is generally to keep a consistent code style across the project. For this project, the designated one chosen is the [airbnb style guide](https://github.com/airbnb/javascript). Its _mostly_ a preference thing but a lot of people really like Airbnb's specifications. 

Now we use Babel to _transpile_ our code to older more commonly supported versions of javascript so that we can write fancy things using newer versions of JS while still making sure all platforms understand our code. 

To follow the Airbnb style guide we need to follow this rule [object-shorthand](https://eslint.org/docs/rules/object-shorthand)

so instead of what you have, we can write: 

```
const job = new CronJob({
    cronTime: `0 0 20 * * *`,
    async onTick() {
      const data = await contribution(store.get('username'));
      if (data.currentStreak === 0 && Notification.isSupported()) {
        new Notification({
          title: 'Streaker',
          body: 'You didn\'t contribute today',
          icon: icon.icon
        }).show()
      }
    }
  });
```

Its a small change which keeps that one line of code simpler.

And because we have babel, it will probably get compiled down to what you wrote originally anyway 😛

If you run `npm run test` in the root dir it will give you a list of the all style errors you have. You can also probably configure vscode to do it automatically.

Sorry this was a long winded answer to a simple question... hope this cleared at least something up.

KeziahMoselle replied:
Thank's a lot for the explanation !

I knew about Airbnb style guide and the utility of Babel but i've never used Babel before and Airbnb style guide too (Used StandardJS)

I will do that ! Thank's again for being kind with a noob like me

yursky commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+      store.set('notification.hours', hours);
+      store.set('notification.minutes', minutes);
+      log.info(`Store updated - time=${time} hours=${hours} minutes=${minutes}`);
+      job.setTime(new CronTime(`0 ${minutes} ${hours} * * *`));
```
The reason its not working is because job.setTime stops running the cronjob here. If we look at the source for node-cron we can see how setTime works. [setTime](https://github.com/kelektiv/node-cron/blob/25b8289b866aaebb14457ca5f2fa826bf679b59a/lib/cron.js#L535)

KeziahMoselle replied:
Oh yeah, thank's ! When I wanted to check the definition with VSCode it redirected me on the `.d.ts` file and I was a bit lost because I don't know TS

KeziahMoselle replied:
When I debug with `console.log(job.nextDates())` and `console.log(job.running)`it returns the correct time and `true` for `job.running` But the notification won't appear

KeziahMoselle replied:
Okay, my bad I was in "Don't disturb" mode on my OS !

yursky replied:
I'll take a look at this tomorrow if you still haven't figured it out. It's late for me haha

yursky replied:
Oh haha ok good

yursky commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+  
+  if (store.get('notification.isEnabled')) {
+    job.start();
+    job.setTime(new CronTime(`0 ${store.get('notification.minutes')} ${store.get('notification.hours')} * * *`));
```
Same here, because we call `setTime()` after we start the time it actually stops it right after

yursky reviewed (changes requested):
Added some comments explaining why its not working and a few style changes.

Commit:
Update src/app.js

Commit:
Fix ESLint

jamieweavis commented:
Thanks for reviewing this @yursky! ❤️ I'm currently at work so I haven't had time to review yet. I will try and get this reviewed and merged within the next few hours.

I'd just like to clarify that this project isn't actually using Babel. The beauty of Electron is that the main process is running on a recent version of Node.js while the renderer processes are running on a recent version of Chrome which both support most ES6/7 features out of the box without need for transpilation!

KeziahMoselle commented:
On MacOS the icon looks like that now : 
![preview.png](https://i.imgur.com/F5NWxt9.png)
In the documentation of Electron : [Notification](https://electronjs.org/docs/api/notification) there is no way to remove the icon at the right and replace the icon at the left ?

jamieweavis commented:
@KeziahMoselle that's because the notification uses the executable icon. If you build the app and then trigger a notification then it will use the correct icon.

That also means you can remove the icon property from the `Notification`, that controls the icon which displays on the right if I recall correctly.

KeziahMoselle commented:
That's right, thank's !
![preview.png](https://i.imgur.com/LXtx4Ei.png)

Commit:
Fix icon

jamieweavis commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+      if (data.currentStreak === 0 && Notification.isSupported()) {
+        new Notification({
+          title: 'Streaker',
+          body: 'You didn\'t contribute today'
```
I think that this would read better as:
> You haven't contributed today

jamieweavis commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+      const data = await contribution(store.get('username'));
+      if (data.currentStreak === 0 && Notification.isSupported()) {
+        new Notification({
+          title: 'Streaker',
```
I don't forsee the project name ever changing however, I've avoided hardcoding the name anywhere in the app in favour of getting it from the package json, i.e. `pjson.name`

jamieweavis commented on src/app.js:
```diff
@@ -170,6 +231,38 @@ app.on('ready', () => {
+  }
+
+  const job = new CronJob({
+    cronTime: `0 0 20 * * *`,
```
The template string here is redundant, can we use single quotes instead?

KeziahMoselle replied:
I forgot to replace it, sorry

jamieweavis replied:
Don't be sorry, it happens 😄

jamieweavis reviewed (changes requested):
Since the notification uses the executable icon, the icons added in this PR can now be safely removed.

Commit:
Update notifications body message

Commit:
Add pjson.name & Replace ` by `

Commit:
Remove notification icons .png

jamieweavis commented:
Great work @KeziahMoselle, and thanks @yursky! 🎉

KeziahMoselle commented:
Thank's ! I will create the preferences window soon 🎉
